staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go

```diff
@@ -270,21 +289,73 @@ func (s *Validator) validateExpressions(ctx context.Context, fldPath *field.Path
 						return errs, -1
 					} else {
 						klog.V(2).ErrorS(msgErr, "messageExpression evaluation failed")
-						errs = append(errs, field.Invalid(fldPath, sts.Type, ruleMessageOrDefault(rule)))
+						errs = append(errs, returnErrBasedOnReason(fldPath, sts.Type, ruleMessageOrDefault(rule), rule.Reason))
 						remainingBudget = newRemainingBudget
 					}
 				} else {
-					errs = append(errs, field.Invalid(fldPath, sts.Type, messageExpression))
+					errs = append(errs, returnErrBasedOnReason(fldPath, sts.Type, messageExpression, rule.Reason))
 					remainingBudget = newRemainingBudget
 				}
 			} else {
-				errs = append(errs, field.Invalid(fldPath, sts.Type, ruleMessageOrDefault(rule)))
+				errs = append(errs, returnErrBasedOnReason(fldPath, sts.Type, ruleMessageOrDefault(rule), rule.Reason))
 			}
 		}
 	}
 	return errs, remainingBudget
 }
 
+func returnErrBasedOnReason(fldPath *field.Path, value interface{}, detail string, reason *metav1.StatusReason) *field.Error {
```

Comment:
nit: don't put "return" in the name of functions. Maybe name this function `fieldErrorForReason` ?